### ISSUE TITLE
Adopt the official jsdoc way of documenting nullable/non-nullable types?

### ISSUE THREAD

**Issue by zepumph** [phetsims/tasks#935]
I was poking around the jsdoc documentation for https://github.com/phetsims/phet-io-wrappers/issues/74 and found the following segment about annotating types for parameters and such:

http://usejsdoc.org/tags-type.html

![image](https://user-images.githubusercontent.com/6856943/39607069-58d0b3d2-4ee5-11e8-8d0e-ec680c48e238.png)

Currently I see 707 usages in the project of the regex `\{.*\|{0,1}null[\}\|]`, which I think gets us pretty close to the real thing. I'm not purposing that we go back and fix everything immediately, but I like having this option, it is a bit simpler. 

Tagging for dev meeting to get thoughts and consensus.

**Comment by zepumph** [comment]
I would be inclined to agree. My vote would be to support `{?number}` to mean number number or null.

**Comment by pixelzoom** [comment]
If we're going to start following the spec...  The spec also indicates `{(number|boolean)}` where we've been using `{number|boolean}`.

**Comment by jonathanolson** [comment]
I guess if you have a more complicated expression, you'd need to wrap it, e.g. `{number|boolean|null}` => `?(number|boolean)`? We would also have `Property.<?number>` (that would be valid?)

I feel strongly that the default "nothing specified" should be non-nullable.

I don't have a strong preference otherwise. I anticipate a lot of cases that are tricky to update, and cases that I won't immediately read as easily (`{?Array.<?Node>}`), but will probably be easier to read once I'm used to it.

**Comment by zepumph** [comment]
I know there are a lot of pieces of the jsdoc spec that we don't touch, and I'm not saying we should. Currently it doesn't seems like `{(number|boolean)}` is providing any benefit over `{number|boolean}`. Specifically for the nullable annotation, I do see an improvement. `{?Array.<?Node>}` over `{Array.<Node|null>|null}`. It is a single character that means the same thing as 5, and imo improves readability.

**Comment by zepumph** [comment]
When annotating nullable for a arg that can take multiple types, I think it would be annotated as 
`{?number|boolean|Array.<string>}`

not

`{?(number|boolean|Array.<string>)}`

**Comment by jonathanolson** [comment]
> When annotating nullable for a arg that can take multiple types, I think it would be annotated as
`{?number|boolean|Array.<string>}`

Sounds good to me.

Presuming we go with this, would we just update documentation as we see it, or would we try to do global replacements?

**Comment by zepumph** [comment]
I'm not very interested in a global find/replace. I would rather chip away at it.

**Comment by jbphet** [comment]
I'm up for using the question mark to mean nullable type, and not using the exclamation point for non-nullable (seems like unnecessary noise).

**Comment by zepumph** [comment]
Above is a commit where I converted `Accessibility.js` to use this, I thought it could help illustrate some of the different usages, and it is easily revertible.

**Comment by samreid** [comment]
TLDR: Usage of `?` in Accessibility.js seems reasonable to me.  

Longer explanation: The versions with `|null` is slightly easier for me to read at the moment, but I suspect that's because I'm more accustomed to it at the moment and I'll get used to `?` quickly.  Not a big deal one way or the other to me, but I do prefer heading in the direction of standards instead of something we made up because we weren't aware of the standard.

**Comment by zepumph** [comment]
We talked about this more today, and @pixelzoom brought up a good point that because we currently aren't planning to adopt parens around multiple types like `{(number|string}}`, so we aren't using legal jsdoc anyways by adopting a solution like `{?number|string}`.

I don't feel too strongly, and I'm beginning to feel like this change is just muddying clear enough waters. There are some places for PhET-iO where it would be really nice to use this annotation, since we are actually rendering jsdoc, but otherwise I don't largely care (that is in about 8 files, mostly in phet-io repos). 

We will keep talking about this, but for now i'll revert the changes to the a11y trait.

**Comment by pixelzoom** [comment]
If we're not concerned with being JSdoc-compliant, and we consider both `null` and `undefined`, I prefer `{string|null|undefined}` to something like `{?string|undefined}` or `{?(string|undefined)}`.  I'm not a fan of using special notation for special cases.  But I'll follow the consensus.

**Comment by zepumph** [comment]
It has been decided that we will not adopt using a `?` to document nullable types generally. In the few places for phet-io docs that require this to be properly rendered in jsdoc, we will keep it in there, and be ok with this minor discrepancy.  If in the future, we feel like this needs to be revisited, then we will. Closing

**Comment by customcommander** [comment]
Hi,

Total stranger here! I struggled with this myself and found this thread while I was doing a bit of research on the subject and thought I'd share back what I've found.

It was a surprise to me but the type expressions `{number}` and `{Number}` aren't the same at all. One describe a literal which is non-nullable by default, the other describe an instance which is nullable by default. Therefore `{number}` is the same as `{!number}` and `{Number}` is the same as `{?Number}`.

This can be checked with the Google Closure Compiler which uses JSDoc types expressions:

```javascript
/** @param {Number} x */
const a = x => x;

/** @param {number} x */
const b = x => x;

a();               b();               // ERR: a & b
a(undefined);      b(undefined);      // ERR: a & b
a(null);           b(null);           // ERR: b
a(42);             b(42);             // ERR: a
a(new Number(42)); b(new Number(42)); // ERR: b
```

See fiddle [here](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%2520%2540param%2520%257BNumber%257D%2520x%2520*%252F%250Aconst%2520a%2520%253D%2520x%2520%253D%253E%2520x%253B%250A%250A%252F**%2520%2540param%2520%257Bnumber%257D%2520x%2520*%252F%250Aconst%2520b%2520%253D%2520x%2520%253D%253E%2520x%253B%250A%250Aa()%253B%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520b()%253B%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%252F%252F%2520ERR%253A%2520a%2520%2526%2520b%250Aa(undefined)%253B%2520%2520%2520%2520%2520%2520b(undefined)%253B%2520%2520%2520%2520%2520%2520%252F%252F%2520ERR%253A%2520a%2520%2526%2520b%250Aa(null)%253B%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520b(null)%253B%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%252F%252F%2520ERR%253A%2520b%250Aa(42)%253B%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520b(42)%253B%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%252F%252F%2520ERR%253A%2520a%250Aa(new%2520Number(42))%253B%2520b(new%2520Number(42))%253B%2520%252F%252F%2520ERR%253A%2520b%250A).

> We talked about this more today, and @pixelzoom brought up a good point that because we currently aren't planning to adopt parens around multiple types like {(number|string}}, so we aren't using legal jsdoc anyways by adopting a solution like {?number|string}.

I don't think that's true.

JSDoc is able to find error in type expression. Say you have this in _example.js_

```javascript
/** @param {?nonsense?} x */
const a = x => x;
```

When you run `yarn -s jsdoc example.js` an error is displayed:

```
ERROR: Unable to parse a tag's type expression for source file...
```

Doing the same with this code is fine:

```javascript
/** @param {?string|number} x */
const a = x => x;
```

And we can see it still holds true with GCC:

```javascript
/** @param {?string|number} x */
const a = x => x;

a();          // ERR (arg count)
a(null);      // OK
a(undefined); // ERR (type mismatch; can be null but not undefined)
a("42");      // OK
a(42);        // OK
```

See fiddle [here](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%2520%2540param%2520%257B%253Fstring%257Cnumber%257D%2520x%2520*%252F%250Aconst%2520a%2520%253D%2520x%2520%253D%253E%2520x%253B%250A%250Aa()%253B%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%252F%252F%2520ERR%2520(arg%2520count)%250Aa(null)%253B%2520%2520%2520%2520%2520%2520%252F%252F%2520OK%250Aa(undefined)%253B%2520%252F%252F%2520ERR%2520(type%2520mismatch%253B%2520can%2520be%2520null%2520but%2520not%2520undefined)%250Aa(%252242%2522)%253B%2520%2520%2520%2520%2520%2520%252F%252F%2520OK%250Aa(42)%253B%2520%2520%2520%2520%2520%2520%2520%2520%252F%252F%2520OK%250A).

**Conclusion**

1. `{number}` is non-nullable by default. 
2. If you mean a number or null then you must use `{?number}`
3. If you mean a number or boolean or null then `{?number|boolean}` is appropriate

If you're interested I have more details here
https://gist.github.com/customcommander/5206dcb44670e34f6923b62c1781e1d2

**Comment by samreid** [comment]
I tested the following code in the google closure compiler:

```js
/** @param {number|null} x */
const a = x => x;

a(7);
a(null);
```
and the compiler said it was acceptable:

![image](https://user-images.githubusercontent.com/679486/103492648-82ac9800-4de9-11eb-9193-340e50360f32.png)


https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%2520%2540param%2520%257Bnumber%257Cnull%257D%2520x%2520*%252F%250Aconst%2520a%2520%253D%2520x%2520%253D%253E%2520x%253B%250A%250Aa(7)%253B%250Aa(null)%253B

**Comment by customcommander** [comment]
Absolutely this is fine too. `{?number}` is just syntactic sugar for `{number|null}`.

> Shorthand for the union of the null type with any other type. This is just syntactic sugar.

<img width="936" alt="Screenshot 2021-01-04 at 01 42 40" src="https://user-images.githubusercontent.com/681975/103494313-2e77d700-4e2e-11eb-9ae0-816f7169818f.png">

See https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System

**Comment by codekandis** [comment]
It should be noted that there is a distinct meaning if one uses  `?`, `!` or none of both if in combination with any type.

Lets have a look on the `*` type.

| Type | Meaning |
| - | - |
| * | Can be a value of any type **INCLUDING** `null`. `null` does not have any specific meaning. It's just any of the possible values. |
| ?* | Can be a value of any type **OR** `null`. `null` is not a possible value but has a specific meaning. This must be documented. |
| !* | Can be a value of any type but **NOT** `null`. |

An example for `?*` - assumed on arrays whose prohibit nullable element values.

```javascript
/**
 * Represents the callback of any array iteration.
 * @callback Array_IteratedElementCallback
 * @param {!*} element The currently iterated element.
 * @param {!Number} index The index of the currently iterated element.
 */

/**
 * Finds the first element specified by a predicate.
 * @param {Array_IteratedElementCallback} predicate The predicate to determine the element.
 * @return {!*} The element, if found, null otherwise.
 */
Array.prototype.findFirstOrNull( predicate )
{
	const foundElement = this.find( predicate );

	return undefined === foundElement
		? null
		: foundElement;
}
```

So it's clear in this example, that the return value is `any` but NOT `null`, while `null` means no element has been found.

Why do not return `undefined` like `Array.find()` does? `undefined` is a possible result from various errors or nonexistant members / variables etc., while `null` is defined by code and more precise.
